subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt.c

```diff
@@ -354,44 +398,186 @@ static int fs_mgmt_file_upload(struct smp_streamer *ctxt)
 		if (len == ULLONG_MAX) {
 			return MGMT_ERR_EINVAL;
 		}
+	}
 
-		fs_mgmt_ctxt.uploading = true;
-		fs_mgmt_ctxt.off = 0;
-		fs_mgmt_ctxt.len = len;
-	} else {
-		if (!fs_mgmt_ctxt.uploading) {
-			return MGMT_ERR_EINVAL;
+	if (k_sem_take(&fs_mgmt_ctxt.lock_sem, FILE_SEMAPHORE_MAX_TAKE_TIME)) {
+		return MGMT_ERR_EBUSY;
+	}
+
+	/* Check if there is an active upload in progress */
+	if (ctxt->smpt != fs_mgmt_ctxt.transport) {
+		/* Different transport, clean up */
+		close_file = true;
+		open_file = true;
+	} else if (fs_mgmt_ctxt.state == STATE_UPLOAD) {
+		/* Check if this is the same file */
+		if (strcmp(file_name, fs_mgmt_ctxt.path)) {
+			/* Different file, we need to clean up */
+			close_file = true;
+			open_file = true;
 		}
+	} else if (fs_mgmt_ctxt.state == STATE_DOWNLOAD) {
+		/* Different mode, we need to clean up */
+		close_file = true;
+		open_file = true;
+	} else {
+		open_file = true;
+	}
```

Comment:
```
if (ctxt->smpt != fs_mgmt_ctxt.transport || fs_mgmt_ctxt.state != this_state || strcmp(file_name, fs_mgmt_ctxt.path))
{
    fs_mgmt_cleanup(); // Cleanup seems to be safe regardless current state;
    /* Open file stuff here as whatever is going to happen file needs to be repoened, because cleanup closed it in aboveline */
}
```